src/v/kafka/server/write_at_offset_stm.cc

```diff
@@ -81,24 +83,33 @@ raft::replicate_stages write_at_offset_stm::replicate(
     return raft::replicate_stages{
       std::move(f),
       do_replicate(
-        std::move(batch),
-        expected_base_offset,
+        std::move(batches),
+        std::move(expected_base_offsets),
         prev_log_offset,
         timeout,
         std::move(as),
         std::move(enqueued_promise))};
 }
 
 ss::future<result<raft::replicate_result>> write_at_offset_stm::do_replicate(
-  model::record_batch batch,
-  kafka::offset expected_base_offset,
+  chunked_vector<model::record_batch> batches,
+  chunked_vector<kafka::offset> expected_base_offsets,
   std::optional<kafka::offset> prev_log_offset,
   model::timeout_clock::duration timeout,
   std::optional<std::reference_wrapper<ss::abort_source>> as,
   ss::promise<> enqueued_promise) {
+    if (batches.empty() || expected_base_offsets.size() != batches.size())
```

Comment:
nit: either here or in the loop below we should validate `expected_base_offsets` is an increasing sequence of offsets, and that the `batches` should should be as well. As it stands there is extra validation that happens only on the first element of both of these vectors.

Reply:
done.